carddav/server.go

```diff
@@ -614,43 +615,160 @@ func (b *backend) propFindAllAddressObjects(ctx context.Context, propfind *inter
 }
 
 func (b *backend) PropPatch(r *http.Request, update *internal.PropertyUpdate) (*internal.Response, error) {
-	homeSetPath, err := b.Backend.AddressBookHomeSetPath(r.Context())
-	if err != nil {
-		return nil, err
-	}
-
+	resType := b.resourceTypeAtPath(r.URL.Path)
 	resp := internal.NewOKResponse(r.URL.Path)
 
-	if r.URL.Path == homeSetPath {
-		// TODO: support PROPPATCH for address books
+	switch resType {
+	case resourceTypeAddressBook:
+		ab, err := b.Backend.GetAddressBook(r.Context(), r.URL.Path)
+		if err != nil {
+			return nil, err
+		}
+		err = b.propPatchAddressBook(r.Context(), update, ab, resp)
+		if err != nil {
+			return nil, err
+		}
+		err = b.Backend.UpdateAddressBook(r.Context(), ab)
+		if err != nil {
+			return nil, err
+		}
+	case resourceTypeAddressObject:
+		dataReq := AddressDataRequest{AllProp: true}
+		ao, err := b.Backend.GetAddressObject(r.Context(), r.URL.Path, &dataReq)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: support PROPPATCH for address objects
+		err = b.propPatchAddressObject(r.Context(), update, ao, resp)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: interface for updating contacts?
+		//err = b.Backend.UpdateAddressObject(r.Context(), ab)
+		//if err != nil {
+		//	return nil, err
+		//}
+	default:
 		for _, prop := range update.Remove {
-			emptyVal := internal.NewRawXMLElement(prop.Prop.XMLName, nil, nil)
-			if err := resp.EncodeProp(http.StatusNotImplemented, emptyVal); err != nil {
-				return nil, err
+			for _, raw := range prop.Prop.Raw {
+				rxn, ok := raw.XMLName()
+				if !ok {
+					return nil, fmt.Errorf("failed to parse properties")
+				}
+				emptyVal := internal.NewRawXMLElement(rxn, nil, nil)
+				if err := resp.EncodeProp(http.StatusMethodNotAllowed, emptyVal); err != nil {
+					return nil, err
+				}
 			}
 		}
 		for _, prop := range update.Set {
-			emptyVal := internal.NewRawXMLElement(prop.Prop.XMLName, nil, nil)
-			if err := resp.EncodeProp(http.StatusNotImplemented, emptyVal); err != nil {
-				return nil, err
+			for _, raw := range prop.Prop.Raw {
+				rxn, ok := raw.XMLName()
+				if !ok {
+					return nil, fmt.Errorf("failed to parse properties")
+				}
+				emptyVal := internal.NewRawXMLElement(rxn, nil, nil)
+				if err := resp.EncodeProp(http.StatusMethodNotAllowed, emptyVal); err != nil {
+					return nil, err
+				}
 			}
 		}
```

Comment:
Do we have to write out a multi-status here? Can't we just reply with a toplevel HTTP status instead? Something like `internal.HTTPErrorf`?